src/fabric/executor/multi_threaded/executor.rs

```diff
@@ -0,0 +1,288 @@
+    /// An index of waiters for incomplete results
+    waiters: HashMap<ResultId, Vec<ResultWaiter<C>>>,
+    /// The thread pool that results are computed on
+    pool: ThreadPool,
+    /// The network outbound queue
+    network_outbound: KanalSender<NetworkOutbound<C>>,
+}
+
+impl<C: CurveGroup> ParallelExecutor<C> {
+    /// Constructor
+    pub fn new(
+        circuit_size_hint: usize,
+        job_queue: Arc<SegQueue<ExecutorMessage<C>>>,
+        network_outbound: KanalSender<NetworkOutbound<C>>,
+    ) -> Self {
+        let pool =
+            ThreadPoolBuilder::new().num_threads(1).build().expect("error building thread pool");
+        Self {
+            job_queue,
+            operations: GrowableBuffer::new(circuit_size_hint),
+            dependencies: GrowableBuffer::new(circuit_size_hint),
+            results: ParallelResultBuffer::new(circuit_size_hint),
+            ready_mask: ResultMask::new(circuit_size_hint),
+            waiters: HashMap::new(),
+            pool,
+            network_outbound,
+        }
+    }
+
+    /// Run the executor until a shutdown message is received
+    pub fn run(mut self) {
+        loop {
+            if let Some(job) = self.job_queue.pop() {
+                match job {
+                    ExecutorMessage::Result(res) => self.handle_new_result(res),
+                    ExecutorMessage::ResultBatch(res) => {
+                        for result in res.into_iter() {
+                            self.handle_new_result(result);
+                        }
+                    },
+                    ExecutorMessage::ResultsReady(ids) => {
+                        for result_id in ids.into_iter() {
+                            self.handle_new_result_by_id(result_id);
+                        }
+                    },
+                    ExecutorMessage::Op(operation) => self.handle_new_operation(operation),
+                    ExecutorMessage::NewWaiter(waiter) => self.handle_new_waiter(waiter),
+                    ExecutorMessage::Shutdown => {
+                        log::debug!("executor shutting down");
+                        break;
+                    },
+                }
+            }
+        }
+    }
+
+    // ----------------
+    // | Job Handlers |
+    // ----------------
+
+    /// Handle a new result
+    fn handle_new_result(&mut self, result: OpResult<C>) {
+        let id = result.id;
+        self.ready_mask.mark_ready(id);
+
+        // Add the result to the buffer and execute any operations that are now ready
+        self.insert_result(result);
+        self.execute_dependents(id);
+    }
+
+    /// Handle a new result given by ID
+    fn handle_new_result_by_id(&mut self, id: ResultId) {
+        // Notify any threads waiting on this value
+        self.wake_waiters_on_result(id);
+
+        self.ready_mask.mark_ready(id);
+        self.execute_dependents(id);
+    }
+
+    /// Handle a new operation
+    fn handle_new_operation(&mut self, mut op: Operation<C>) {
```

Comment:
this method is structured as it is to avoid cloning where possible, hence the lack of control flow unification